src/Lykke.Job.CandleHistoryWriter.Repositories/Candles/AssetPairCandlesHistoryRepository.cs

```diff
@@ -0,0 +1,146 @@
+
+            var candleByRowsChunks = candles
+                .GroupBy(candle => CandleHistoryEntity.GenerateRowKey(candle.Timestamp, _timeInterval))
+                .ToChunks(100);
+
+            foreach (var candleByRowsChunk in candleByRowsChunks)
+            {
+                // If we can't store the candles, we can't do anything else, so just retries until success
+                await Policy
+                    .Handle<Exception>()
+                    .WaitAndRetryForeverAsync(
+                        retryAttempt => TimeSpan.FromSeconds(Math.Pow(2, retryAttempt)),
+                        (exception, timeSpan) =>
+                        {
+                            var context = $"{_assetPairId}-{priceType}-{_timeInterval}";
+
+                            return _log.WriteErrorAsync("Persist candle rows chunk with retries", context, exception);
+                        })
+                    .ExecuteAsync(() => SaveCandlesBatchAsync(candleByRowsChunk, partitionKey));
+            }
+        }
+
+        private async Task SaveCandlesBatchAsync(IEnumerable<IGrouping<string, ICandle>> candleByRowsChunk, string partitionKey)
+        {
+            var candleByRows = candleByRowsChunk.ToDictionary(g => g.Key, g => g.AsEnumerable());
+
+            // updates existing entities
+
+            var existingEntities = (await _tableStorage.GetDataAsync(partitionKey, candleByRows.Keys)).ToArray();
+
+            foreach (var entity in existingEntities)
+            {
+                entity.MergeCandles(candleByRows[entity.RowKey], _assetPairId, _timeInterval);
+            }
+
+            // creates new entities
+
+            var newEntityKeys = candleByRows.Keys.Except(existingEntities.Select(e => e.RowKey));
+            var newEntities = newEntityKeys.Select(k => new CandleHistoryEntity(partitionKey, k)).ToArray();
+
+            foreach (var entity in newEntities)
+            {
+                entity.MergeCandles(candleByRows[entity.RowKey], _assetPairId, _timeInterval);
+            }
+
+            // save changes
+
+            _healthService.TraceCandleRowsPersisted(existingEntities.Length + newEntities.Length);
+
+            await _tableStorage.InsertOrReplaceBatchAsync(existingEntities.Concat(newEntities));
+        }
+
+        public async Task<IEnumerable<ICandle>> GetCandlesAsync(CandlePriceType priceType, CandleTimeInterval interval, DateTime from, DateTime to)
+        {
+            if (priceType == CandlePriceType.Unspecified)
+            {
+                throw new ArgumentException(nameof(priceType));
+            }
+
+            var query = GetTableQuery(priceType, interval, from, to);
+            var entities = await _tableStorage.WhereAsync(query);
+            var candles = entities
+                .SelectMany(e => e.Candles.Select(ci => ci.ToCandle(_assetPairId, e.PriceType, e.DateTime, interval)));
+
+            return candles.Where(c => c.Timestamp >= from && c.Timestamp < to);
```

Comment:
Is it possible to filter on the Azure server side in a query?

Reply:
This is a collection of the candles serialized as JSON in the single Azure Table field